include/oneapi/dpl/internal/scan_by_segment_impl.h

```diff
@@ -0,0 +1,506 @@
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#ifndef _ONEDPL_SCAN_BY_SEGMENT_IMPL_H
+#define _ONEDPL_SCAN_BY_SEGMENT_IMPL_H
+
+#if _ONEDPL_BACKEND_SYCL
+
+#    include <oneapi/dpl/pstl/algorithm_fwd.h>
+#    include <oneapi/dpl/pstl/parallel_backend.h>
+#    include <oneapi/dpl/pstl/hetero/utils_hetero.h>
+
+#    include <oneapi/dpl/pstl/hetero/dpcpp/utils_ranges_sycl.h>
+#    include <oneapi/dpl/pstl/hetero/dpcpp/unseq_backend_sycl.h>
+#    include <oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h>
+
+#    include <array>
+
+namespace oneapi
+{
+namespace dpl
+{
+namespace internal
+{
+
+template <typename _NdItem, typename _LocalAcc, typename _IdxType, typename _ValueType, typename _BinaryOp>
+inline _ValueType
+wg_segmented_scan(_NdItem __item, _LocalAcc __local_acc, _IdxType __local_id, _IdxType __delta_local_id,
+                  _ValueType __accumulator, _BinaryOp __binary_op, ::std::size_t __wgroup_size)
+{
+    _IdxType __first = 0;
+    __local_acc[__local_id] = __accumulator;
+
+    sycl::group_barrier(__item.get_group());
+
+    for (int32_t __i = 1; __i < __wgroup_size; __i += __i)
+    {
+        if (__delta_local_id >= __i)
+            __accumulator = __binary_op(__local_acc[__first + __local_id - __i], __accumulator);
+
+        __first = __wgroup_size - __first;
+        __local_acc[__first + __local_id] = __accumulator;
+
+        sycl::group_barrier(__item.get_group());
+    }
+
+    return (__local_id ? __local_acc[__first + __local_id - 1] : 0);
+}
+
+enum class scan_type
+{
+    inclusive,
+    exclusive
+};
+
+template <scan_type __scan_type>
+struct sycl_scan_by_segment_impl
+{
+    template <typename _ExecutionPolicy, typename _Range1, typename _Range2, typename _Range3,
+              typename _BinaryPredicate, typename _BinaryOperator>
+    void
+    sycl_inclusive_scan_by_segment(_ExecutionPolicy&& __exec, _Range1&& __keys, _Range2&& __values,
+                                   _Range3&& __out_values, _BinaryPredicate __binary_pred, _BinaryOperator __binary_op)
+    {
+        using __diff_type = oneapi::dpl::__internal::__difference_t<_Range1>;
+        using __key_type = oneapi::dpl::__internal::__value_t<_Range1>;
+        using __val_type = oneapi::dpl::__internal::__value_t<_Range2>;
+        using __flag_type = bool;
+
+        const int __n = __keys.size();
+
+        constexpr int __vals_per_item = 2; // Assigning 2 elements per work item resulted in best performance on gpu.
+
+        ::std::size_t __wgroup_size = oneapi::dpl::__internal::__max_work_group_size(__exec);
+
+        // change __wgroup_size according to local memory limit
+        __wgroup_size = oneapi::dpl::__internal::__max_local_allocation_size(
+            ::std::forward<_ExecutionPolicy>(__exec), sizeof(__key_type) + sizeof(__val_type), __wgroup_size);
+
+        int32_t __n_groups = 1 + std::ceil(__n / (__wgroup_size * __vals_per_item));
```

Comment:
Why int32_t ? I think that the size of __n_groups is not large then the size of __n